docs/CONTRIBUTING.md

```diff
@@ -35,3 +35,47 @@ Test is determined fragile/broken when it is failing despite no defects are pres
   - Test can still be resurrected from history if needed - in ATH or plugin itself.
 
 The situation is expected to be handled in 90 days since reported. In case of no corrective action, the test will be removed from ATH.
+
+## Guidelines to avoid fragile/flaky tests
+
+### Click and wait
+
+When writing acceptance tests using Selenium it is common to click on an element and then run some other action on the resulting page.
+Most of the time the page is updated really quick (specially if there is not a network request involved), but on slow infrastructure this can take some time so that the next test action after clicking could be checking on something which is not the expected.
+
+To avoid this, use the click and wait pattern:
+
+```
+WebElement clickableElement = ...;
+clickableElement.click();
+waitFor(by.id("my-element")).withTimeout(10);
+... now run actions safely on my-element ...
+```
+
+This way the test will not be affected by an underlying slow infrastructure.
+
+### Do not rely on external infra
+
+When the test requires to interact with an external system, use a `DockerFixture`.
+There are multiple examples into the [`fixtures` directory](https://github.com/jenkinsci/acceptance-test-harness/tree/master/src/main/java/org/jenkinsci/test/acceptance/docker/fixtures).
+
+### Do not use sleep
+
+Sleeping blindly is a bad practice, it just makes tests execution time longer. 
+Use the [`Wait` class](https://github.com/jenkinsci/acceptance-test-harness/blob/master/src/main/java/org/jenkinsci/test/acceptance/junit/Wait.java) as much as possible. `CapybaraPortingLayerImpl#waitFor` is an easy way to create instances of `Wait`.
+
+### Use robust selectors
+
+When selecting elements from the DOM, use robust selectors. 
+In general, try to avoid selectors that are known to change or are outside your control (they are not in your code). 
+
+A few examples:
+
+1. Do not use full xpath expressions. DOM can change anytime.
+2. Avoid using parts of the DOM which are not under your control (a change in core or any other plugin can make your test to fail).
```

Comment:
I disagree on this one.
Using ID is a bad practice, specially in a plugable system. IDs must be unique in the DOM, two elements with the same ID can cause a lot of problems.

IDs should be avoided.

Reply:
Well, defining an ID is something that someone chooses carefully, a collision will rarely occur. Otherwise we have problems to find the element in the DOM in the code under test as well.  Using complex selectors is also quite fragile (see 1.) - in the test code *and* in the code under test. So I understand this is a risk, but it is a price I am willing to pay in order to simplify things.

Reply:
So this implies we need a section on what makes a good id (prefixed with the plugin-id? for example to avoid any collisions).
Id's don;t work well on f:repeatable for example.

Reply:
We are talking about two different things here:

- Using hardcoded ids on elements that can repeat on a page (ex.: shell build step), which is not a good idea, obviously.
- Ids accidentally colliding with something unrelated on the page.

As Antonio said, ids would not help in the first case and that is where `path` attribute comes to rescue. Though are we really suffering from the latter?

Reply:
>  Though are we really suffering from the latter

I do not know as I am generally not using IDs.  this was in response to @uhafner's comment.

it sounds like we are saying "do not use IDs for selecting elements" but Ulli was saying the opposite?

Reply:
Added a note about this.